core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -47,55 +47,79 @@ private[server] class MetadataCache(brokerId: Int) extends Logging {
-                case e: Throwable =>
-                  debug("Error while fetching metadata for %s: %s".format(topicPartition, e.toString))
-                  new PartitionMetadata(partitionId, leaderInfo, replicaInfo, isrInfo,
-                    Errors.forException(e).code)
-              }
+  private def getPartitionMetadata(topic: String, protocol: SecurityProtocol): Option[Iterable[MetadataResponse.PartitionMetadata]] = {
+    cache.get(topic).map { partitions =>
+      partitions.map { case (partitionId, partitionState) =>
+        val topicPartition = TopicAndPartition(topic, partitionId)
+
+        val leaderAndIsr = partitionState.leaderIsrAndControllerEpoch.leaderAndIsr
+        val leader = aliveBrokers.get(leaderAndIsr.leader).map(_.getNode(protocol))
+
+        val replicas = partitionState.allReplicas
+        val replicaInfo = getAliveEndpoints(replicas, protocol)
+
+        if (leader.isEmpty) {
+          debug("Error while fetching metadata for %s: leader not available".format(topicPartition))
+          new MetadataResponse.PartitionMetadata(Errors.LEADER_NOT_AVAILABLE, partitionId, Node.noNode(),
+            replicaInfo.asJava, java.util.Collections.emptyList())
+        } else {
+          val isr = leaderAndIsr.isr
+          val isrInfo = getAliveEndpoints(isr, protocol)
+
+          if (replicaInfo.size < replicas.size) {
+            debug("Error while fetching metadata for %s: replica information not available for following brokers %s"
+              .format(topicPartition, replicas.filterNot(replicaInfo.map(_.id).contains(_)).mkString(",")))
+
+            new MetadataResponse.PartitionMetadata(Errors.REPLICA_NOT_AVAILABLE, partitionId, leader.get,
+              replicaInfo.asJava, isrInfo.asJava)
+          } else if (isrInfo.size < isr.size) {
+            debug("Error while fetching metadata for %s: in sync replica information not available for following brokers %s"
+              .format(topicPartition, isr.filterNot(isrInfo.map(_.id).contains(_)).mkString(",")))
+            new MetadataResponse.PartitionMetadata(Errors.REPLICA_NOT_AVAILABLE, partitionId, leader.get,
+              replicaInfo.asJava, isrInfo.asJava)
+          } else {
+            new MetadataResponse.PartitionMetadata(Errors.NONE, partitionId, leader.get, replicaInfo.asJava,
+              isrInfo.asJava)
           }
-          topicResponses += new TopicMetadata(topic, partitionMetadata.toSeq)
         }
       }
     }
-    topicResponses
   }
 
-  def getAliveBrokers = {
+  def getTopicMetadata(topics: Set[String], protocol: SecurityProtocol): Seq[MetadataResponse.TopicMetadata] = {
+    inReadLock(partitionMetadataLock) {
+      val topicsRequested = if (topics.isEmpty) cache.keySet else topics
+      topicsRequested.flatMap { topic =>
+        getPartitionMetadata(topic, protocol).map { partitionMetadata =>
+          new MetadataResponse.TopicMetadata(Errors.NONE, topic, partitionMetadata.toBuffer.asJava)
+        }
+      }.toSeq
```

Comment:
Nice refactoring. One minor improvement that you can do is to do the `toSeq` before the `flatMap` instead of after. That way you avoid adding everything to a `Set` (which checks for duplicates and is more expensive).